### PR TITLE
feature: add MultiSigTransactionHAMT to MultiSig actor

### PR DESCRIPTION
**Pull request by frrist**

### CODE REVIEW

File: actors/builtin/multisig/multisig_actor.go [hunk 1]
```diff
@@ -303,13 +306,16 @@ func (a *MultiSigActor) _rtApproveTransactionOrAbort(rt Runtime, callerAddr addr
 			txn.Params,
 			txn.Value,
 		)
+		// TODO REVIEW if getPendingTxn returns the hamt node we can save a look up by passing it here can deleting.
```
Comment (frrist) [inline actors/builtin/multisig/multisig_actor.go]: engrish: meant to say we can save a look up by passing the hamt node to the delete method for slightly faster look up when deleting.

File: actors/builtin/multisig/multisig_actor.go [hunk 2]
```diff
@@ -120,8 +117,14 @@ func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) TxnID {
 		Value:      params.Value,
 	}
 
-	st.PendingTxns[txnID] = txn
 	st.PendingApprovals[txnID] = autil.ActorIDSetHAMT_Empty()
+
+	pendingTxnCID, err := setPendingTxn(context.TODO(), rt, txnID, txn)
+	if err != nil {
+		rt.AbortStateMsg(fmt.Sprintf("failed to set transaction in HAMT: %v", err))
+	}
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Push the error check and abort down into the helper method. And for getPendingTxn and deletePendingTxn too

File: actors/builtin/multisig/multisig_actor.go [hunk 3]
```diff
@@ -120,8 +117,14 @@ func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) TxnID {
 		Value:      params.Value,
 	}
 
-	st.PendingTxns[txnID] = txn
 	st.PendingApprovals[txnID] = autil.ActorIDSetHAMT_Empty()
+
+	pendingTxnCID, err := setPendingTxn(context.TODO(), rt, txnID, txn)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: The name pendingTxnCID is confusing. This is assigning over the HAMT root node, so should be `newPendingTxnsRoot` or similar.

File: actors/builtin/multisig/multisig_actor.go [hunk 4]
```diff
@@ -335,3 +341,44 @@ func UpdateRelease_MultiSig(rt Runtime, h vmr.ActorStateHandle, st MultiSigActor
...
+func deletePendingTxn(ctx context.Context, rt vmr.Runtime, rcid cid.Cid, txnID TxnID) error {
+	nd, err := hamt.LoadNode(ctx, vmr.RtCborwrapper{rt}, rcid)
+	if err != nil {
+		return err
+	}
+	return nd.Delete(ctx, t2k(txnID))
+}
+
+// convert a TxnID to a HAMT key.
+func t2k(txnID TxnID) string {
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: This name is too obtuse. Try `asKey()`, `toKey()`, `key()` ...

File: actors/runtime/runtime_util.go [hunk 5]
```diff
@@ -78,3 +80,18 @@ func RT_ConfirmFundsReceiptOrAbort_RefundRemainder(rt Runtime, fundsRequired abi
 		rt.SendFunds(rt.ImmediateCaller(), rt.ValueReceived()-fundsRequired)
 	}
 }
+
+type RtCborwrapper struct {
```
[the remark this answers is not in the document]
Comment (anorth) [reply actors/runtime/runtime_util.go]: - Unexport the struct and name it something like `store`, `cborStore` or similar
- Export a concisely named method like `asStore(rt)`
- Add a static type assertion for the new interface
- Put a comment in the code, not in the PR

File: actors/util/actor_util.go [hunk 6]
```diff
@@ -1,12 +1,31 @@
...
+func init() {
+	bs := bstore.NewBlockstore(dstore.NewMapDatastore())
+	cst := hamt.CSTFromBstore(bs)
+	nd := hamt.NewNode(cst)
+	emptyHAMT, err := cst.Put(context.TODO(), nd)
+	if err != nil {
+		panic(err)
+	}
+	EmptyHAMT = emptyHAMT
+}
```
Comment (anorth) [inline actors/util/actor_util.go]: Whoa! Let's reign in those dependencies. We certainly don't want any of the stores in order to store nothing.

File: go.mod [hunk 7]
```diff
@@ -5,6 +5,9 @@ go 1.13
 require (
 	github.com/filecoin-project/go-address v0.0.0-20200107215422-da8eea2842b5
 	github.com/ipfs/go-cid v0.0.4
+	github.com/ipfs/go-datastore v0.3.1
+	github.com/ipfs/go-hamt-ipld v0.0.14
+	github.com/ipfs/go-ipfs-blockstore v0.1.1
```
Comment (anorth) [inline go.mod]: Only the hamt library is acceptable here.

File: actors/util/actor_util.go (hunk 6 above)
Comment (frrist) [reply actors/util/actor_util.go]: Is the above better?

**Pull request merged**